src/Ast/BaseNode.php

```diff
@@ -0,0 +1,37 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\PhpDocParser\Ast;
+
+abstract class BaseNode implements Node
+{
+
+	/** @var array<string, mixed> */
+	private $attributes = [];
```

Comment:
Why not just `public $attributes = []` without the methods? All other properties on nodes are public as well.

Reply:
I like the convenience methods `get`/`set`/`has`, with public property you'd have juggle with ugly `isset` or `array_key_exists` in userland code...

Reply:
See https://github.com/phpstan/phpdoc-parser/pull/57/files/1647b310b0d93d2b070875d96558e721335caa07#diff-1c468e54ea94bf2ee8fdc56af68ed6c184a4a54c3298ab74baad9b6404c98bf1

Also better type declaration support on method types.
Making it `public` would override everything with assigns, which is not desired on attributes (compared to all other properties on nodes)